### PR TITLE
Do not set up subscription when setting a delegate for a MTRDevice wi…

### LINKED ISSUE #33557: Do not set up subscription for remote controllers when delegate is set
Currently when we set delegate on a MTRDevice, we always set up subscriptions by default. However for remote controllers (over XPC) we should not be setting up subscriptions.

### PR DESCRIPTION
**Pull request by nivi-apple**
…th a remote controller over XPC

- Add tests for testing the behavior when a MTRDevice is created with an XPC controller

Fixes: #33557

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTRDevice.mm [hunk 1]
```diff
@@ -689,7 +689,9 @@ - (void)setDelegate:(id<MTRDeviceDelegate>)delegate queue:(dispatch_queue_t)queu
 {
     MTR_LOG_INFO("%@ setDelegate %@", self, delegate);
 
-    BOOL setUpSubscription = YES;
+    // We should not set up a subscription for device controllers over XPC.
+    BOOL setUpSubscription = ![_deviceController isKindOfClass:MTRDeviceControllerOverXPC.class];
+    ;
```
Comment (jtung-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:694]: Extraneous `;`

Comment (nivi-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:694]: :)

**Review by woody-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTRDevice.mm [hunk 2]
```diff
@@ -685,11 +685,17 @@ - (void)_setDSTOffsets:(NSArray<MTRTimeSynchronizationClusterDSTOffsetStruct *>
...
+{
+    return ![_deviceController isKindOfClass:MTRDeviceControllerOverXPC.class];
+}
+
 - (void)setDelegate:(id<MTRDeviceDelegate>)delegate queue:(dispatch_queue_t)queue
 {
     MTR_LOG_INFO("%@ setDelegate %@", self, delegate);
 
-    BOOL setUpSubscription = YES;
+    // We should not set up a subscription for device controllers over XPC.
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:697]: Shouldn't this comment be in _subscriptionsAllowed?

File: src/darwin/Framework/CHIP/MTRDevice.mm [hunk 3]
```diff
@@ -685,11 +685,17 @@ - (void)_setDSTOffsets:(NSArray<MTRTimeSynchronizationClusterDSTOffsetStruct *>
 #define MTR_DEVICE_SUBSCRIPTION_MAX_INTERVAL_MIN (10 * 60) // 10 minutes (for now)
 #define MTR_DEVICE_SUBSCRIPTION_MAX_INTERVAL_MAX (60 * 60) // 60 minutes
 
+- (BOOL)_subscriptionsAllowed
+{
+    return ![_deviceController isKindOfClass:MTRDeviceControllerOverXPC.class];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:690]: Shouldn't we use this helper in the other place where we are doing this check?

File: src/darwin/Framework/CHIP/MTRDevice.mm [hunk 4]
```diff
@@ -685,11 +685,17 @@ - (void)_setDSTOffsets:(NSArray<MTRTimeSynchronizationClusterDSTOffsetStruct *>
...
+}
+
 - (void)setDelegate:(id<MTRDeviceDelegate>)delegate queue:(dispatch_queue_t)queue
 {
     MTR_LOG_INFO("%@ setDelegate %@", self, delegate);
 
-    BOOL setUpSubscription = YES;
+    // We should not set up a subscription for device controllers over XPC.
+    BOOL setUpSubscription = [self _subscriptionsAllowed];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:698]: This is using `_deviceController` without locking.  That might be OK, but have you checked that?

File: src/darwin/Framework/CHIPTests/MTRDeviceTests.m [hunk 5]
```diff
@@ -3618,6 +3618,36 @@ - (void)test034_TestMTRDeviceHistoricalEvents
...
+    __auto_type * delegate = [[MTRDeviceTestDelegate alloc] init];
+
+    XCTAssertTrue([device _getInternalState] == MTRInternalDeviceStateUnsubscribed);
+
+    delegate.onAttributeDataReceived = ^(NSArray<NSDictionary<NSString *, id> *> * attributeReport) {
+        [subscriptionExpectation fulfill];
+    };
+
+    [device setDelegate:delegate queue:queue];
+    [self waitForExpectations:@[ subscriptionExpectation ] timeout:30];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/MTRDeviceTests.m:3646]: So this test will always take 30 seconds to run?  That is pretty painful, in terms of actually running the tests locally....  Why do we need this bit at all, given the internal state check above?

File: src/darwin/Framework/CHIPTests/MTRDeviceTests.m [hunk 6]
```diff
@@ -3618,6 +3618,36 @@ - (void)test034_TestMTRDeviceHistoricalEvents
...
+    __auto_type * device = [MTRDevice deviceWithNodeID:kDeviceId deviceController:remoteController];
+    dispatch_queue_t queue = dispatch_get_main_queue();
+
+    // We should not set up a subscription when creating a MTRDevice with a remote controller.
+    XCTestExpectation * subscriptionExpectation = [self expectationWithDescription:@"Subscription has been set up"];
+    subscriptionExpectation.inverted = YES;
+
+    __auto_type * delegate = [[MTRDeviceTestDelegate alloc] init];
+
+    XCTAssertTrue([device _getInternalState] == MTRInternalDeviceStateUnsubscribed);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/MTRDeviceTests.m:3639]: Why isn't this XCTAssertEqual?

File: src/darwin/Framework/CHIPTests/TestHelpers/MTRTestDeclarations.h [hunk 7]
```diff
@@ -59,10 +59,31 @@ NS_ASSUME_NONNULL_BEGIN
 @end
 
 @interface MTRDevice (TestDebug)
+typedef NS_ENUM(NSUInteger, MTRInternalDeviceState) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/TestHelpers/MTRTestDeclarations.h:62]: Why is this here?  Seems like we should just move the "internal device state" to a header that both MTRDevice.mm and the tests include instead of duplicating the enum and having problems when our duplication gets out of sync.

Comment (nivi-apple) [reply src/darwin/Framework/CHIPTests/TestHelpers/MTRTestDeclarations.h:62]: done.

File: src/darwin/Framework/CHIPTests/MTRDeviceTests.m (hunk 6 above)
Comment (nivi-apple) [reply src/darwin/Framework/CHIPTests/MTRDeviceTests.m:3639]: done.

File: src/darwin/Framework/CHIPTests/MTRDeviceTests.m (hunk 5 above)
Comment (nivi-apple) [reply src/darwin/Framework/CHIPTests/MTRDeviceTests.m:3646]: i wanted to wait a few seconds before checking the internal state again and also make sure the inverted exception is not called. would 5 secs be ok or you are saying we do not need to wait just check the state after setting delegate?

File: src/darwin/Framework/CHIP/MTRDevice.mm (hunk 2 above)
Comment (nivi-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:697]: done.

File: src/darwin/Framework/CHIP/MTRDevice.mm (hunk 3 above)
Comment (nivi-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:690]: done.
